crates/block-producer/src/block_producer.rs

```diff
@@ -258,17 +258,20 @@ impl BlockProducer {
                 let db = self.store.begin_transaction();
                 let block_smt = db.reverted_block_smt()?;
 
-                let local_root: &H256 = block_smt.root();
+                let local_root: H256 = (*block_smt.root()).into();
                 let global_revert_block_root: H256 = global_state.reverted_block_root().unpack();
-                assert_eq!(local_root, &global_revert_block_root);
+                assert_eq!(local_root, global_revert_block_root);
 
                 let keys: Vec<H256> = collected_block_hashes.into_iter().collect();
                 for key in keys.iter() {
                     log::info!("submit revert block {:?}", hex::encode(key.as_slice()));
                 }
-                let proof = block_smt
-                    .merkle_proof(keys.clone())?
-                    .compile(keys.clone())?;
+                let proof = {
+                    let smt_keys: Vec<_> = keys.iter().map(|k| (*k).into()).collect();
```

Comment:
The compiler *should* be able to optimize this to a no-op:

https://stackoverflow.com/questions/48308759/how-do-i-convert-a-vect-to-a-vecu-without-copying-the-vector

Without `repr(transparent)` on SMT's H256, it's not really safe to skip this conversion with unsafe code.

Reply:
Actually, I just tried it and the optimization is much more likely to kick in when using `into_iter`. Consider using `into_iter` or construct a `Vec<SMTH256>` in the first place.